module/zfs/zio.c

```diff
@@ -4864,16 +4872,38 @@ zio_checksum_verify(zio_t *zio)
 		ASSERT3U(zio->io_prop.zp_checksum, ==, ZIO_CHECKSUM_LABEL);
 	}
 
+	ASSERT0(zio->io_flags & ZIO_FLAG_DIO_CHKSUM_ERR);
+
 	if ((error = zio_checksum_error(zio, &info)) != 0) {
 		zio->io_error = error;
 		if (error == ECKSUM &&
 		    !(zio->io_flags & ZIO_FLAG_SPECULATIVE)) {
-			mutex_enter(&zio->io_vd->vdev_stat_lock);
-			zio->io_vd->vdev_stat.vs_checksum_errors++;
-			mutex_exit(&zio->io_vd->vdev_stat_lock);
-			(void) zfs_ereport_start_checksum(zio->io_spa,
-			    zio->io_vd, &zio->io_bookmark, zio,
-			    zio->io_offset, zio->io_size, &info);
+			if (zio->io_flags & ZIO_FLAG_DIO_READ) {
+				zio->io_flags |= ZIO_FLAG_DIO_CHKSUM_ERR;
```

Comment:
It seems for speculative read (`ZIO_FLAG_SPECULATIVE`) you won't set `ZIO_FLAG_DIO_CHKSUM_ERR` here, that I guess may cause recovery writes on higher levels. It seems this `if` should move moved up.

Reply:
I looked over the code, and I am not aware of anyway that a Direct I/O read could be issued as speculative. The only place I spotted as maybe a concern was data prefetching but that does not happen with Direct I/O reads. I did add an `IMPLY` statement to `zio_checksum_verify()` to make sure this is never the case though.